### PR TITLE
Fix gpu halo update on cpu

### PR DESCRIPTION
**Pull request by eddie-c-davis**
This PR makes some (admittedly somewhat hacky) fixes to the `feature/gpu_halo_update_on_cpu` to get the halo exchange tests passing on Piz Daint.

```bash
srun --partition=debug --account=s1053 -N 6 -C gpu pytest -vv -s tests/mpi/test_mpi_halo_update.py --gpu-only
srun: job 28729408 queued and waiting for resources
srun: job 28729408 has been allocated resources
...
======================= 216 passed, 360 skipped in 6.42s =======================
tests/mpi/test_mpi_halo_update.py::test_zeros_vector_halo_update[same-3-interface_3d-gt4py_cupy-1] PASSED

======================= 216 passed, 360 skipped in 6.47s =======================
tests/mpi/test_mpi_halo_update.py::test_zeros_vector_halo_update[same-3-interface_3d-gt4py_cupy-1] PASSED

======================= 216 passed, 360 skipped in 6.03s =======================
tests/mpi/test_mpi_halo_update.py::test_zeros_vector_halo_update[same-3-interface_3d-gt4py_cupy-1] PASSED

======================= 216 passed, 360 skipped in 6.49s =======================
tests/mpi/test_mpi_halo_update.py::test_zeros_vector_halo_update[same-3-interface_3d-gt4py_cupy-1] PASSED

======================= 216 passed, 360 skipped in 6.49s =======================
tests/mpi/test_mpi_halo_update.py::test_zeros_vector_halo_update[same-3-interface_3d-gt4py_cupy-1] PASSED

======================= 216 passed, 360 skipped in 6.50s =======================

```

### CODE REVIEW

File: fv3gfs/util/buffer.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from ._timing import Timer, NullTimer
 from numpy import ndarray
 import contextlib
+import cupy as cp
```
Comment (mcgibbon) [inline fv3gfs/util/buffer.py:5]: Need to contain this in a try: except ModuleNotFoundError block, with the appropriate change to logic below
```suggestion
try:
    import cupy as cp
except ModuleNotFoundError:
    cp = None
```

File: tests/mpi/test_mpi_halo_update.py [hunk 2]
```diff
@@ -245,7 +245,8 @@ def depth_quantity(
 ):
     """A quantity whose value indicates the distance from the computational
     domain boundary."""
-    data = numpy.zeros(shape, dtype=dtype) + numpy.nan
+    data = numpy.zeros(shape, dtype=dtype)
+    data[:] = numpy.nan
```
Comment (mcgibbon) [inline tests/mpi/test_mpi_halo_update.py:249]: At some point we'll catch all of these.

File: fv3gfs/util/buffer.py [hunk 3]
```diff
@@ -91,4 +92,4 @@ def recv_buffer(allocator: Callable, array: ndarray, timer: Optional[Timer] = No
             timer.stop("unpack")
             yield recvbuf
             with timer.clock("unpack"):
-                array[:] = recvbuf
+                array[:] = cp.asarray(recvbuf) if isinstance(array, cp.ndarray) else recvbuf
```
Comment (mcgibbon) [inline fv3gfs/util/buffer.py:95]: Please add a code comment explaining why this was important, edit as needed.
```suggestion
                # without cp.asarray, it throws a CUDA error about
                # memory ownership on the cupy arrays from the gt4py storages
                # this should be fixed in a later version of gt4py
                array[:] = cp.asarray(recvbuf) if isinstance(array, cp.ndarray) else recvbuf
```

**Review by mcgibbon — APPROVED** [review]
Some changes needed, noted below. I suggested two of them, the last one is adding appropriate logic to handle the case where `cp` is `None` if you merge one of the suggestions. Approving anyways since it's straightforward and this is getting merged into another feature branch.

**Comment by mcgibbon** [comment]
Also seems to be failing lint checks at the moment.

**Comment by eddie-c-davis** [comment]
launch jenkins

File: fv3gfs/util/utils.py [hunk 4]
```diff
@@ -50,3 +55,18 @@ def is_c_contiguous(array: Union[np.ndarray, Storage]) -> bool:
 def ensure_contiguous(maybe_array: Union[np.ndarray, Storage]) -> None:
     if isinstance(maybe_array, np.ndarray) and not is_contiguous(maybe_array):
         raise ValueError("ndarray is not contiguous")
+
+
+def assign_array(
+    left_array: Union[np.ndarray, Storage], right_array: Union[np.ndarray, Storage],
```
Comment (mcgibbon) [inline fv3gfs/util/utils.py:61]: Please rename these `to_array` and `from_array`, it's not as obvious these refer to the "left" and "right" sides of an equation.

File: fv3gfs/util/utils.py [hunk 5]
```diff
@@ -50,3 +55,18 @@ def is_c_contiguous(array: Union[np.ndarray, Storage]) -> bool:
 def ensure_contiguous(maybe_array: Union[np.ndarray, Storage]) -> None:
     if isinstance(maybe_array, np.ndarray) and not is_contiguous(maybe_array):
         raise ValueError("ndarray is not contiguous")
+
+
+def assign_array(
```
Comment (mcgibbon) [inline fv3gfs/util/utils.py:60]: Great idea collecting these into a central place!

**Review by mcgibbon — APPROVED** [review]
Great!

**Pull request merged**